File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/indexes/IndexedExample.cs
using System;
using System.IO;
using com.db4o;
using com.db4o.query;

namespace com.db4odoc.f1.indexes
{	
	public class IndexedExample
	{
		public readonly static string YapFileName = "formula1.yap";	

		public static void NoIndex() {
    		ObjectContainer db=Db4o.OpenFile(YapFileName);
			try {
    			Query query = db.Query();
				query.Constrain(typeof(Car));
				query.Descend("_pilot").Descend("_points").Constrain("99");

				DateTime dt1 = DateTime.UtcNow;
				ObjectSet  result = query.Execute();
				DateTime dt2 = DateTime.UtcNow;
				TimeSpan  diff = dt2 - dt1;
				Console.WriteLine("Test 1: no indexes");
				Console.WriteLine("Execution time="+diff.Milliseconds + " ms");
				ListResult(result);
			}
			finally {
				db.Close();
			}
		}
		// end NoIndex
	    
		public static void FillUpDB(){
			File.Delete(YapFileName);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try {
        		for (int i=0; i<10000;i++){
    				AddCar(db,i);
    			}
			}
			finally {
				db.Close();
			}
		}
		// end FillUpDB
	  
		public static void PilotIndex() {
    		Db4o.Configure().ObjectClass(typeof(Car)).ObjectField("_pilot").Indexed(true);
    		Db4o.Configure().ObjectClass(typeof(Pilot)).ObjectField("_points").Indexed(false);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try {
    			Query query = db.Query();
				query.Constrain(typeof(Car));
				query.Descend("_pilot").Descend("_points").Constrain("99");

				DateTime dt1 = DateTime.UtcNow;
				ObjectSet  result = query.Execute();
				DateTime dt2 = DateTime.UtcNow;
				TimeSpan  diff = dt2 - dt1;
				Console.WriteLine("Test 3: index on pilot");
				Console.WriteLine("Execution time="+diff.Milliseconds + " ms");
				ListResult(result);
			}
			finally {
				db.Close();
			}
		}
		// end PilotIndex
	   
		public static void PointsIndex() {
    		Db4o.Configure().ObjectClass(typeof(Car)).ObjectField("_pilot").Indexed(false);
    		Db4o.Configure().ObjectClass(typeof(Pilot)).ObjectField("_points").Indexed(true);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try {
    			Query query = db.Query();
				query.Constrain(typeof(Car));
				query.Descend("_pilot").Descend("_points").Constrain("99");

				DateTime dt1 = DateTime.UtcNow;
				ObjectSet  result = query.Execute();
				DateTime dt2 = DateTime.UtcNow;
				TimeSpan  diff = dt2 - dt1;
				Console.WriteLine("Test 4: index on points");
				Console.WriteLine("Execution time="+diff.Milliseconds + " ms");
				ListResult(result);
			}
			finally {
				db.Close();
			}
		}
	    // end PointsIndex
	    
		public static void FullIndex() {
    		Db4o.Configure().ObjectClass(typeof(Car)).ObjectField("_pilot").Indexed(true);
    		Db4o.Configure().ObjectClass(typeof(Pilot)).ObjectField("_points").Indexed(true);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try {
    			Query query = db.Query();
				query.Constrain(typeof(Car));
				query.Descend("_pilot").Descend("_points").Constrain("99");

				DateTime dt1 = DateTime.UtcNow;
				ObjectSet  result = query.Execute();
				DateTime dt2 = DateTime.UtcNow;
				TimeSpan  diff = dt2 - dt1;
				Console.WriteLine("Test 2: index on pilot and points");
				Console.WriteLine("Execution time="+diff.Milliseconds + " ms");
				ListResult(result);
			}
			finally {
				db.Close();
			}
		}
		// end FullIndex
	    
		private static void AddCar(ObjectContainer db, int points)
		{
			Car car = new Car("BMW");
			car.Pilot= new Pilot("Tester", points);
			db.Set(car);
		}
		// end AddCar
	    
		public static void ListResult(ObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/indexes/IndexedExample.vb
Imports System
Imports System.IO
Imports com.db4o
Imports com.db4o.query

Namespace com.db4odoc.f1.indexes
    Public Class IndexedExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"

        Public Shared Sub NoIndex()
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                query.Descend("_pilot").Descend("_points").Constrain("99")
                Dim dt1 As DateTime = DateTime.UtcNow
                Dim result As ObjectSet = query.Execute()
                Dim dt2 As DateTime = DateTime.UtcNow
                Dim diff As TimeSpan = dt2 - dt1
                Console.WriteLine("Test 1: no indexes")
                Console.WriteLine("Execution time=" + diff.Milliseconds.ToString() + " ms")
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
        ' end NoIndex

        Public Shared Sub FillUpDB()
            File.Delete(YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                For i As Integer = 0 To 10000
                    AddCar(db, i)
                Next
            Finally
                db.Close()
            End Try
        End Sub
        ' end FillUpDB

        Public Shared Sub PilotIndex()
            Db4o.Configure().ObjectClass(GetType(Car)).ObjectField("_pilot").Indexed(True)
            Db4o.Configure().ObjectClass(GetType(Pilot)).ObjectField("_points").Indexed(False)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                query.Descend("_pilot").Descend("_points").Constrain("99")
                Dim dt1 As DateTime = DateTime.UtcNow
                Dim result As ObjectSet = query.Execute()
                Dim dt2 As DateTime = DateTime.UtcNow
                Dim diff As TimeSpan = dt2 - dt1
                Console.WriteLine("Test 3: index on pilot")
                Console.WriteLine("Execution time=" + diff.Milliseconds.ToString() + " ms")
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
        ' end PilotIndex

        Public Shared Sub PointsIndex()
            Db4o.Configure().ObjectClass(GetType(Car)).ObjectField("_pilot").Indexed(False)
            Db4o.Configure().ObjectClass(GetType(Pilot)).ObjectField("_points").Indexed(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                query.Descend("_pilot").Descend("_points").Constrain("99")
                Dim dt1 As DateTime = DateTime.UtcNow
                Dim result As ObjectSet = query.Execute()
                Dim dt2 As DateTime = DateTime.UtcNow
                Dim diff As TimeSpan = dt2 - dt1
                Console.WriteLine("Test 4: index on points")
                Console.WriteLine("Execution time=" + diff.Milliseconds.ToString() + " ms")
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
        ' end PointsIndex


        Public Shared Sub FullIndex()
            Db4o.Configure().ObjectClass(GetType(Car)).ObjectField("_pilot").Indexed(True)
            Db4o.Configure().ObjectClass(GetType(Pilot)).ObjectField("_points").Indexed(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                query.Descend("_pilot").Descend("_points").Constrain("99")
                Dim dt1 As DateTime = DateTime.UtcNow
                Dim result As ObjectSet = query.Execute()
                Dim dt2 As DateTime = DateTime.UtcNow
                Dim diff As TimeSpan = dt2 - dt1
                Console.WriteLine("Test 2: index on pilot and points")
                Console.WriteLine("Execution time=" + diff.Milliseconds.ToString() + " ms")
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
        ' end FullIndex


        Public Shared Sub AddCar(ByVal db As ObjectContainer, ByVal points As Integer)
            Dim car As Car = New Car("BMW")
            Dim pilot As Pilot = New Pilot("Tester", points)
            car.Pilot = pilot
            db.[Set](car)
        End Sub
        ' end AddCar

        Public Shared Sub ListResult(ByVal result As ObjectSet)
            Console.WriteLine(result.Count)
            For Each item As Object In result
                Console.WriteLine(item)
            Next
        End Sub
        ' end ListResult
	End Class
End Namespace